\chapter{Introduction}\label{introduction}
The aim of the project is to discover disease subtypes of Prostate adenocarcinoma from a dataset coming from The Cancer Genome Atlas (TCGA)~\cite{hutter2018TCGA}. So we take into consideration the dataset with code ``\textit{PRAD}". In particular, we will consider the subtypes identified by The Cancer Genome Atlas Research Network~\cite{abeshouse2015molecularPRAD}. We will be looking at a subset of the types of data analyzed, namely the biomolecules obtained from mRNA and miRNA sequencing, and reverse-phase protein array of the samples. Indeed the study takes into account information characterized through whole-exome sequencing for somatic mutations, array-based methods for profiling both somatic copy-number changes and DNA methylation, mRNA and microRNA (miRNA) sequencing, reverse-phase protein array (RPPA), and low-pass and high-pass whole-genome sequencing (WGS). Only for a few samples non-malignant adjacent prostate samples were also examined for DNA methylation and RNA/miRNA expression analyses.

The Cancer Genome Atlas Research Network performed unsupervised clustering of data from each molecular platform, as well as integrative clustering using iCluster~\cite{shen2009integrative}. The latter approach is more ideal since it allows joint inference from multi-omics data by generating a single integrated cluster assignment by simultaneously capturing patterns of genomic alterations. Instead, identifying tumor subtypes by separately clustering each data view requires manual integration of the results. The analyses with both individual and integrative clustering of the TCGA Prostate adenocarcinoma dataset uncovered both known and novel associations, with 74\% of all tumors being assignable to one of seven molecular classes. These are based on distinct oncogenic drivers: fusions involving (1) ERG, (2) ETV1, (3) ETV4, or (4) FLI1; mutations in (5) SPOP
or (6) FOXA1; or (7) IDH1 mutations. Thus, more generally, the tumors can be subdivided into 3 disease subtypes. We will try to predict the membership of a sample in the \textit{PRAD} multi-omics dataset to a class by applying varying integration methods and different clustering algorithms.\newline

Practically speaking we pre-process the multi-omics TCGA dataset, as described in Section~\ref{methods_preProcessing}. This results in a dataset of only a subset of the original samples, where the values of the features have been normalized. Seen that we are working with different data views for the same samples it is necessary to integrate the data, so to obtain a unique object that summarizes the distance between all considered samples. We use the integration data approaches presented in Section~\ref{methods_dataInt}. This information is used as input for clustering algorithms, with the aim of classifying the samples into one of the three disease subtypes. This phase is detailed in Section~\ref{methods_clustering}.

Finally, we present the classification results for each type of integration method, combined with all chosen clustering techniques. We show the obtained clusterings visually. More importantly we evaluate the quality of the results by comparing with appropriate indices the subtypes assigned to each sample in the project with the subdivision given by iCluster.